include/aws/s3/private/aws_mmap.h

```diff
@@ -0,0 +1,62 @@
+
+#ifndef AWS_MMAP_H
+#define AWS_MMAP_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+/**
+ * TODO: move this to part of <aws/common/file.h> instead
+ */
+
+#include <aws/s3/s3.h>
+
+struct aws_allocator;
+
+AWS_EXTERN_C_BEGIN
+
+struct aws_mmap_context;
+
+/**
+ * TODO: supports different mode and extra.
+ * Create a READ-ONLY mmap context.
+ */
+AWS_S3_API struct aws_mmap_context *aws_mmap_context_new(struct aws_allocator *allocator, const char *file_name);
+
+/**
+ * TODO: real refcount.
+ */
+AWS_S3_API struct aws_mmap_context *aws_mmap_context_release(struct aws_mmap_context *context);
+
+/**
+ * Get mapped content. OS will load the content into memory when the content is accessed.
+ * You need to invoke `aws_mmap_context_unmap_content` on the `out_start_addr` once read from the memory is done,
+ * otherwise, leak will happen
+ *
+ * @param context           The mmap context
+ * @param length            The length of memory to map
+ * @param offset            The offset starting at the file to map.
+ * @param out_start_addr    The start address for the mapped memory, it will be a multiple
+ *                          of the page size as returned by sysconf(_SC_PAGE_SIZE)
```

Comment:
I guess the dream of making the codebase less complex didn't pan out. Nor is it getting any better performance. So all this really brings is that we don't have to worry about running out of file descriptors.

Have the pure-Java transfer managers every run into issues with file-descriptor limits? If it's never really been an issue for them, then maybe we should go that route, for simplicity's sake. We can always pull this mmap stuff out of our back pockets if customers report they're hitting those limits